xmtp_mls/src/groups/device_sync.rs

```diff
@@ -399,540 +358,514 @@ where
+            Err(DeviceSyncError::AlreadyAcknowledged) => {
+                return Ok(());
+            }
+            result => result?,
         }
 
-        Ok(())
-    }
+        let Some(device_sync_server_url) = &self.device_sync.server_url else {
+            tracing::info!("Unable to send sync payload - no sync server url present.");
+            return Err(DeviceSyncError::MissingSyncServerUrl);
+        };
 
-    async fn create_sync_reply(
-        &self,
-        request_id: &str,
-        syncables: &[Vec<Syncable>],
-        kind: DeviceSyncKind,
-    ) -> Result<DeviceSyncReplyProto, DeviceSyncError> {
-        let (payload, enc_key) = encrypt_syncables(syncables)?;
-
-        // upload the payload
-        let Some(url) = &self.history_sync_url else {
-            return Err(DeviceSyncError::MissingHistorySyncUrl);
+        let mut request_id = "".to_string();
+        let options = if let Some(request) = request {
+            let Some(options) = request.options else {
+                return Err(DeviceSyncError::MissingOptions);
+            };
+            request_id = request.request_id;
+            options
+        } else {
+            default_backup_options()
         };
-        let upload_url = format!("{url}/upload");
-        tracing::info!(
-            inbox_id = self.inbox_id(),
-            installation_id = hex::encode(self.installation_public_key()),
-            "Using upload url {upload_url}",
-        );
 
-        let response = reqwest::Client::new()
-            .post(upload_url)
-            .body(payload)
-            .send()
-            .await?;
+        // Generate a random encryption key
+        let key = xmtp_common::rand_vec::<32>();
+
+        // Now we want to create an encrypted stream from our database to the history server.
+        //
+        // 1. Build the exporter
+        let exporter = BackupExporter::new(options, &provider, &key);
+        let metadata = exporter.metadata().clone();
+
+        #[cfg(not(target_arch = "wasm32"))]
+        let body = {
+            // 2. A compat layer to have futures AsyncRead play nice with tokio's AsyncRead
+            let exporter_compat = tokio_util::compat::FuturesAsyncReadCompatExt::compat(exporter);
+            // 3. Add a stream layer over the async read
+            let stream = tokio_util::io::ReaderStream::new(exporter_compat);
+            // 4. Pipe that stream as the body to the request to the history server
+            reqwest::Body::wrap_stream(stream)
+        };
+        #[cfg(target_arch = "wasm32")]
+        let body = {
+            use futures::AsyncReadExt;
+            // Make exporter mutable
+            let mut exporter = exporter;
+
+            // Wasm does not support stream uploads. So we'll just consume the stream into a vec.
+            let mut buffer = Vec::new();
+            exporter.read_to_end(&mut buffer).await?;
+            buffer
+        };
+
+        // 5. Make the request
```

Comment:
these steps are really great. Wondering if it would make sense to split it off into its own Struct type with each step as a function

something like

```
let send = DeviceSyncSend::new(/* .. */);
send.build()
send.stream()
send.request()
send.reply()
```

etc.

This would be similar to modeling as state machine, and has a few benefits:
- easier to test each individual step
- observe state before/after each step for logging etc.
- readability (each function would do only one thing)
- separation of concerns/concepts